haystack/pipelines/config.py

```diff
@@ -0,0 +1,119 @@
+import copy
+import logging
+import os
+from pathlib import Path
+import re
+from typing import Any, Dict, List, Optional
+from networkx import DiGraph
+import yaml
+
+
+logger = logging.getLogger(__name__)
+
+
+VALID_CODE_GEN_INPUT_REGEX = re.compile(r"^[-a-zA-Z0-9_/.:]+$")
+
+
+def get_pipeline_definition(pipeline_config: Dict[str, Any], pipeline_name: Optional[str] = None) -> Dict[str, Any]:
+    """
+    Get the definition of Pipeline from a given pipeline config. If the config contains more than one Pipeline,
+    then the pipeline_name must be supplied.
+
+    :param pipeline_config: Dict Pipeline config parsed as a dictionary.
+    :param pipeline_name: name of the Pipeline.
+    """
+    if pipeline_name is None:
+        if len(pipeline_config["pipelines"]) == 1:
+            pipeline_definition = pipeline_config["pipelines"][0]
+        else:
+            raise Exception("The YAML contains multiple pipelines. Please specify the pipeline name to load.")
+    else:
+        pipelines_in_definitions = list(filter(lambda p: p["name"] == pipeline_name, pipeline_config["pipelines"]))
+        if not pipelines_in_definitions:
+            raise KeyError(f"Cannot find any pipeline with name '{pipeline_name}' declared in the YAML file.")
+        pipeline_definition = pipelines_in_definitions[0]
+
+    return pipeline_definition
+
+
+def get_component_definitions(pipeline_config: Dict[str, Any], overwrite_with_env_variables: bool) -> Dict[str, Any]:
+    """
+    Returns the definitions of all components from a given pipeline config.
+
+    :param pipeline_config: Dict Pipeline config parsed as a dictionary.
+    :param overwrite_with_env_variables: Overwrite the YAML configuration with environment variables. For example,
+                                            to change index name param for an ElasticsearchDocumentStore, an env
+                                            variable 'MYDOCSTORE_PARAMS_INDEX=documents-2021' can be set. Note that an
+                                            `_` sign must be used to specify nested hierarchical properties.
+    """
+    component_definitions = {}  # definitions of each component from the YAML.
+    raw_component_definitions = copy.deepcopy(pipeline_config["components"])
+    for component_definition in raw_component_definitions:
+        if overwrite_with_env_variables:
+            _overwrite_with_env_variables(component_definition)
+        name = component_definition.pop("name")
+        component_definitions[name] = component_definition
+
+    return component_definitions
+
+
+def read_pipeline_config_from_yaml(path: Path):
+    with open(path, "r", encoding="utf-8") as stream:
+        return yaml.safe_load(stream)
+
+
+def validate_config(pipeline_config: Dict[str, Any]):
```

Comment:
Yep, recursion should work here. However I think it's less explicit as it would simply check types for dicts, lists and strings, not really the semantic parts we care about. It should at least check whether there are a components and a pipelines key, whether these entries have name and type keys and so on. On the other hand the recursive approach it would work nicely if we added something.
I'll leave it up to you in the next PR ;-)